### ISSUE TITLE
MIGRATED: WebView does not send custom request header after the first initial page load (#8693)

### ISSUE THREAD

**Issue by jamonholmgren** [react-native-community/react-native-webview#4]
Migrated from React Native core: https://github.com/facebook/react-native/issues/8693

@chiraggshah says:

---

The source prop for WebView allows for passing request headers. When the webview renders for the first time, the request headers are sent.

But when the links within that page are clicked, the headers are not sent to the server in the request header.
Similarly, when the goBack() method is called on the WebView, the request headers are not being sent.

---

_More context available in the original issue._

**Comment by ErrorPro** [comment]
I would also like to have those headers being sent to all the internal resources as well. For example, a page has some static resources(images/css/js) and when those resources are being fetched from the server we might want to pass those headers along with those requests.

**Comment by jamonholmgren** [comment]
@ErrorPro Do you have some time to dig into this and see what it might take to implement?  There are other webview implementations out there that might have already solved this issue.

**Comment by ErrorPro** [comment]
@jamonholmgren Ok, I'll try to dig into it :)

**Comment by jamonholmgren** [comment]
Yay @ErrorPro !!

**Comment by mausworks** [comment]
Related: #18

**Comment by feliperaul** [comment]
We need this so badly! The suggestion of using a handler for each request would be nice, since it accommodates both needs (the ones that need headers only on initial load and the ones that need it on every request). Also, maybe you need different headers on first request (auth token) and other headers on subsequent requests.

**Comment by qubitron** [comment]
For those who need a quick solution they can use right now, I've created something that works for ios by patching the RCTWebView.m control in react-native:

Here's a diff from RN 0.56.0 (this is the version I'm currently using):
https://github.com/qubitron/react-native/commit/d4df41b771b7edb9221da7c6873d1f56abefe8fc

I realized as I was implementing this that I don't want auth headers to be arbitrarily sprayed out to sites accessed by the webview control, so my solution was to use an extra `sendHeadersToHosts` entry in the source property, so that the headers are only sent to hosts that match the desired list.

With the above patch applied, the following works generally:
```
<WebView
    ...
    source={{
        uri: 'foo.com',
        headers: {'name': 'value'},
        sendHeadersToHosts: ['foo.com']
    }}> 
```
Headers are then sent to all GET, POST, etc requests originating from the webview with host name containing foo.com.

I have not yet been successful getting the same approach to work on Android, as it is significantly harder to intercept requests there. 

The WebView APIs on Android make it difficult to make this work for anything other than GET requests. While it's possible to intercept GET, POST requests by overriding
[WebViewClient.shouldInterceptRequest](https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest)) method, the callback does not provide the POST body making it impossible to re-create the desired request.

Custom javascript needs to be injected into the webview to intercept POST calls and get the AJAX/form data body. There is a sample solution for this in the following repo:
https://github.com/KeejOow/android-post-webview

I have not had time to attempt this yet, but it looks like the only approach that might work.

**Comment by pablichjenkov** [comment]
It appears to be a bug in Android Webview recent implementations. By recent I mean Target sdk 24 and up. 
Has anyone reported an issue in the Android bug tracking portal yet?
I am willing to.

**Comment by jpodpro** [comment]
still no solution? i found a work-around for android only, but isn't setting headers a fundamentally important mechanism for mobile apps?
```
onShouldStartLoadWithRequest={(navigator) =>
{
  this.setState({url: navigator.url});
  return false;
}}
```

**Comment by pablichjenkov** [comment]
Not sure if is the same issue you are having. This is the problem I was having with headers not being sent. It was a bug in chrome version 71. Apparently got fixed in version 72.
https://bugs.chromium.org/p/chromium/issues/detail?id=925559#c7

**Comment by tmaly1980** [comment]
@jpodpro Your solution works only for GET requests, not POST requests in case there are form submissions within the WebView.

**Comment by mausworks** [comment]
@harryharihar there is none.

**Comment by sibelius** [comment]
you need to implement `shouldInterceptRequest` on android

**Comment by tonmanayo** [comment]
Is this in the timeline for the future?

**Comment by BettyJJ** [comment]
I come here almost every day to see if it's fixed.  😔

**Comment by punksta** [comment]
I was trying to use `shouldInterceptRequest`. 
Original request implementation works on chromium binary level. Overriding `shouldInterceptRequest` might cause other issues related to webview under the hood things. It seems like `shouldInterceptRequest` is supposed to be used for loading data from custom cache, not for request modification.

**Comment by zuandre** [comment]
Guy's just want to figure out if this issue is being worked on at the moment and if no one figured out a work around.
Initially I ad authorization header to my webview and it gives me logged in user. Navigating around the site breaks and user is logged out.

Is there no way of persisting a header ?

**Comment by mausworks** [comment]
@zuandre No, there is currently no way to persist headers.

That is what this issue is about.

**Comment by jamonholmgren** [comment]
Status: still waiting on a community PR to be submitted to fix this issue.

Workaround: there are a few in this thread, notably: https://github.com/react-native-community/react-native-webview/issues/4#issuecomment-427671845

**Comment by fortydegrees** [comment]
After 3 years has this still not been addressed?

**Comment by connor-odoherty** [comment]
Does anyone have a different workaround from the one suggested above? People using Expo are forced to use a specific version of React Native, and I believe Expo has a non-negligible number of users

**Comment by ArthurBrum** [comment]
Sorry, I'm kinda newbie. But there is this react-native-webview from community same as react-native?
at my package.json Im using:
`"dependencies": {
        ...
       "react-native-webview": "^7.5.2",
       ...
}`

Can I still use solution from 
> #4 (comment)

or should I change to 
`"dependencies": {
        ...
       "@react-native-community/react-native-webview": "^7.5.2",
       ...
}` first?

**Comment by jcohenho** [comment]
Is there any work being done on this? I'm experiencing this exact issue on iOS when trying to load a WebView and then submit a form. I'd be glad to take the work done above from @qubitron and create a PR for the latest version of RN.

**Comment by chiraggshah** [comment]
@jamonholmgren Can this issue be reopened?

**Comment by meshaabi** [comment]
@chiraggshah is your patch from the 2016 article still working ?
https://blog.bigbinary.com/2016/07/26/passing-request-headers-on-each-webview-request-in-react-native.html

I think I just hit this.

**Comment by chiraggshah** [comment]
@meshaabi : Yup.

**Comment by fwon** [comment]
Still No solution for Y-2020 ???!! 
Did react-native-webview updated to solve it ?

**Comment by jamonholmgren** [comment]
Status: Still waiting on a community-submitted solution for this problem.

WebView is maintained by volunteers (with limited free time). We rely on the community to help us devise solutions to common problems like this.

Another idea: if someone is interested in hiring a consulting company to fix this, feel free to get in touch with one. (I own one, but won't link it here due to conflict of interest.) I'm happy to work with them to get this pushed out.

**Comment by BettyJJ** [comment]
> Still waiting on a community-submitted solution for this problem.

Sadly, there was a pull request many months ago, but the mods did not review it and let it be closed. This is really discouraging.

**Comment by softmastx** [comment]
Since 2018!!!   Any solution yet?

**Comment by odinsey** [comment]
no ... chiraggshah opened this issue on 11 Jul 2016
https://github.com/facebook/react-native/issues/8693

**Comment by trnc-ck** [comment]
any update on this?

**Comment by mikecreuzer** [comment]
Should this be reopened?

**Comment by trnc-ck** [comment]
> Should this be reopened?

yes pls ✌🏻

**Comment by teseo** [comment]
> Should this be reopened?

It would be great

**Comment by 100phel** [comment]
Any update on this?

**Comment by Flip32** [comment]
Hey, I have a issue related to that. I'm trying to get user location from de server throw the webView. But when the server asks permission to device, in IOS, only, the msg are not translated as de device settings. For example, my Iphone has Portuguese language but, when a receive the alert from the server in may app, the msg are in English. But in other devices brand, works good. I tried to pass some headers instructions, but not successfully. Does someone knows why, and have some workaround?

**Comment by ettingshausen** [comment]
Should this be reopened again?

**Comment by thijs-qv** [comment]
> Should this be reopened again?

Definitely.

**Comment by qlur** [comment]
Holy moly! Imagine not being able to send a custom request in 2021!

**Comment by anasyo10** [comment]
Is there some forks available that works? :|

**Comment by amiel-dev** [comment]
any update on this?

**Comment by jlrojasVipera** [comment]
any update on this?

**Comment by BettyJJ** [comment]
I have a feeling that this project has been abandoned 😥

**Comment by ou2s** [comment]
Why is this issue closed? Reopen, please.

**Comment by daande** [comment]
We are running into this issue as well can we reopen please?

**Comment by ashishAmz** [comment]
any update?

**Comment by hiepnsx** [comment]
any update?

**Comment by redbar0n** [comment]
@Titozzz just inquiring about the status here, since you mentioned on Sep 26, 2018 that "we are committed to fix this".

@ErrorPro did you find out something when you dug into it?

@jamonholmgren could you reopen this issue, since people are still experiencing the issue (and have been for a while)?

**Comment by nullmicgo** [comment]
just found that react-native-webview has a props called "basicAuthCredential".
Not sure everyone is trying to do it for basic auth or not

https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#basicAuthCredential
![image](https://user-images.githubusercontent.com/1070485/155350789-7468ce90-7cd1-4ce3-9da8-ef7ba4a4f0bc.png)

**Comment by stonebinox** [comment]
Nearly 4 years and counting! We need this reopened! Pretty basic stuff.

**Comment by jasperkuperus** [comment]
As this issue does not get any traction at all and is already closed, I decided to open a new issue, referencing this issue. Hopefully that will attract some attention of the maintainers. This is still an active issue. And I don't feel it will be re-opened soon, so I've opened a new one.

**Comment by YesSkyscrapers** [comment]
I solved this problem for my work app. I created local server inside app, which "proxy" all requests with authroization headers. Inside html I changed all links to "http://localhost:1111/path_to_img". 
For local server used this as template https://github.com/react-native-community-library/react-native-webserver
Its work!

**Comment by hamzahayat** [comment]
I was running into a similar issue. 💀 

I came across the [section](https://github.com/react-native-webview/react-native-webview/blob/master/docs/Guide.md#working-with-custom-headers-sessions-and-cookies) in the `react-native-webview` documentation that helped me solve it!  🙌🏼 

Essentially, you will have to trigger a re-render to make sure that custom headers are injected into every webpage loaded into the `WebView` component. 

## Fix summary
- Storing the `url` in a `React.state` variable 
- Pass a function into the `onShouldStartLoadWithRequest` prop  of the `<WebView />` component.
- This function will be triggered whenever a new request is made in the `WebView` component. 
- In this function you will check whether the new url (available in function params; `request.url`) is the same as the current state url. 
- If it is not, then store the new url in the state, this will trigger a re-render which should inject the custom header in the new page.


## Example from [docs](https://github.com/react-native-webview/react-native-webview/blob/master/docs/Guide.md#working-with-custom-headers-sessions-and-cookies)
```
const CustomHeaderWebView = (props) => {
  const { uri, onLoadStart, ...restProps } = props;
  const [currentURI, setURI] = useState(props.source.uri);
  const newSource = { ...props.source, uri: currentURI };

  return (
    <WebView
      {...restProps}
      source={newSource}
      onShouldStartLoadWithRequest={(request) => {
        // If we're loading the current URI, allow it to load
        if (request.url === currentURI) return true;
        // We're loading a new URL -- change state first
        setURI(request.url);
        return false;
      }}
    />
  );
};

<CustomHeaderWebView
  source={{
    uri: 'http://example.com',
    headers: {
      'my-custom-header-key': 'my-custom-header-value',
    },
  }}
/>;
```

**Comment by jenniredfield** [comment]
Hello. I'd like to post a workaround here too. I had a case where subsequent api calls inside the webview weren't getting the headers either. So a colleague came up with this workaround, where you force the header on the fetch api inside of the webview. 

```
<WebView
  source={{
    headers: {
      ...(props.source.headers ?? {}),
      Authorization: "Basic ***",
    },
  }}
  sharedCookiesEnabled={true}
  webviewDebuggingEnabled={true}
  injectedJavaScriptBeforeContentLoaded={`
    XMLHttpRequest.prototype.open = (function(open) {
      return function(method,url,async) {
        open.apply(this,arguments);
        this.setRequestHeader('Authorization', "Basic ***");
      };
    })(XMLHttpRequest.prototype.open);
    var originalFetch = window.fetch;
    window.fetch = function (input, init) {
        if (!init) {
            init = {};
        }
        if (!init.headers) {
            init.headers = new Headers();
        }
        if (init.headers instanceof Headers) {
            init.headers.append('Authorization', 'Basic ***');
        } else if (init.headers instanceof Array) {
            init.headers.push(['Authorization', 'Basic ***']);
        } else {
            init.headers['Authorization'] = 'Basic ***';
        }
        return originalFetch(input, init);
    };
  `}
  injectedJavaScriptBeforeContentLoadedForMainFrameOnly={false}
  injectedJavaScriptForMainFrameOnly={false}
/>
```

**Comment by VaniaPopovic** [comment]
> Hello. I'd like to post a workaround here too. I had a case where subsequent api calls inside the webview weren't getting the headers either. So a colleague came up with this workaround, where you force the header on the fetch api inside of the webview.
> 
> ```
> <WebView
>   source={{
>     headers: {
>       ...(props.source.headers ?? {}),
>       Authorization: "Basic ***",
>     },
>   }}
>   sharedCookiesEnabled={true}
>   webviewDebuggingEnabled={true}
>   injectedJavaScriptBeforeContentLoaded={`
>     XMLHttpRequest.prototype.open = (function(open) {
>       return function(method,url,async) {
>         open.apply(this,arguments);
>         this.setRequestHeader('Authorization', "Basic ***");
>       };
>     })(XMLHttpRequest.prototype.open);
>     var originalFetch = window.fetch;
>     window.fetch = function (input, init) {
>         if (!init) {
>             init = {};
>         }
>         if (!init.headers) {
>             init.headers = new Headers();
>         }
>         if (init.headers instanceof Headers) {
>             init.headers.append('Authorization', 'Basic ***');
>         } else if (init.headers instanceof Array) {
>             init.headers.push(['Authorization', 'Basic ***']);
>         } else {
>             init.headers['Authorization'] = 'Basic ***';
>         }
>         return originalFetch(input, init);
>     };
>   `}
>   injectedJavaScriptBeforeContentLoadedForMainFrameOnly={false}
>   injectedJavaScriptForMainFrameOnly={false}
> />
> ```

I spent about 10 hours on this one. This is the only solution that ever worked.

**Comment by tdinckaya** [comment]
I’m using React Native 0.76.4, and unfortunately, I experienced this issue with the latest version of react-native-webview. The custom headers were not being sent to the server when the WebView loaded initially. Therefore, I had to downgrade to version 11.22.7. This version works perfectly fine.
